src/gui/toolbarMenubar/model/ColorPalette.cpp

```diff
@@ -0,0 +1,154 @@
+#include "ColorPalette.h"
+
+#include <fstream>
+#include <limits>
+#include <sstream>
+
+#include <gtk/gtk.h>
+#include <util/StringUtils.h>
+
+Palette::Palette(fs::path path): filepath{std::move(path)}, header{}, namedColors{} {}
+
+void Palette::load() {
+    header.clear();
+    namedColors.clear();
+
+    std::ifstream gptFile{filepath};
+    std::string line;
+
+    if (gptFile.is_open()) {
+        getline(gptFile, line);
+        if (!parseFirstGimpPaletteLine(line))
+            return;
+        while (getline(gptFile, line)) {
+            parseHeaderLine(line) || parseColorLine(line);
+        }
+    }
+}
+
+auto Palette::parseFirstGimpPaletteLine(const std::string& line) -> bool {
+    if (StringUtils::trim(line) != "GIMP Palette") {
+        g_error(".gpt file needs to start with \"GIMP Palette\" in the "
+                "first line");
+        return false;
+    }
+    return true;
+}
+
+auto Palette::parseHeaderLine(const std::string& line) -> bool {
+    Header headerEntry;
+    NamedColor color;
+    std::istringstream iss{line};
+    if (iss >> headerEntry) {
+        header[headerEntry.attribute] = headerEntry.value;
+        return true;
+    }
+    return false;
+}
+
+auto Palette::parseColorLine(const std::string& line) -> bool {
+    NamedColor color;
+    std::istringstream iss{line};
+    if (iss >> color) {
+        // call to private variable paletteIndex allowed because of friend class
+        color.paletteIndex = namedColors.size();
+        namedColors.push_back(color);
+        return true;
+    }
+    return false;
+}
+
+
+NamedColor Palette::getColorAt(size_t i) const {
+    if (i >= namedColors.size()) {
+        i = i % namedColors.size();
+        g_warning("There are more Coloritems in the Toolbar then your Palette defines.\n"
+                  "Hence, cycling through palette from the beginning.");
+    }
+
+    return namedColors.at(i);
+}
+
+NamedColor* Palette::getColorPointerAt(size_t i) {
+    if (i >= namedColors.size()) {
+        i = i % namedColors.size();
+        g_warning("There are more Coloritems in the Toolbar then your Palette defines.\n"
+                  "Hence, cycling through palette from the beginning.");
+    }
+
+    return &namedColors.at(i);
+}
+
+size_t Palette::size() const { return namedColors.size(); }
```

Comment:
I would remove those and only return a `std::vector<NamedColor> const&`

Reply:
I think here it makes sense to write the wrapper function. This way the cycling behaviour does not need to be implemented by the caller

```cpp
    if (i >= namedColors.size()) {
        i = i % namedColors.size();
```

Reply:
Only `const NamedColor& Palette::getColorAt(size_t i) const` will be kept.